File: src/test/Ruby/DemoQATest.rb
require 'selenium-webdriver'
require 'test/unit'

# IN THE MIDDLE OF DEBUGGING 


class DemoQATest < Test::Unit::TestCase
  def setup
    options = Selenium::WebDriver::Chrome::Options.new
    # options.add_argument('--headless=new')
    @driver = Selenium::WebDriver.for :chrome, options: options
    @driver.manage.timeouts.implicit_wait = 4 # seconds
    @driver.manage.window.resize_to(1920, 1080)
    @wait = Selenium::WebDriver::Wait.new(timeout: 10) # seconds
  end

  def scroll_past_element(element)
    @driver.switch_to.active_element.send_keys(:page_down)
  end

  def test_demoqa
    username = 'research_paper'
    password = '1tq52rw43E!'

    @driver.navigate.to 'https://demoqa.com/login'
    @driver.find_element(id: 'userName').send_keys(username)
    @driver.find_element(id: 'password').send_keys(password)
    @driver.find_element(id: 'login').click

    store_button = @driver.find_element(id: 'gotoStore')
    scroll_past_element(store_button)
    store_button = @driver.find_element(id: 'gotoStore')
    store_button.click

    @driver.find_element(xpath: "//*[@id='see-book-Git Pocket Guide']").click  # id is supposed to be used here instead of xpath
    book_name = @driver.find_element(xpath: '//*[@id="title-wrapper"]//*[@id="userName-value"]').text # css #title_wrapper #userName-value is supposed to be here
    add_new_record_button = @driver.find_element(id: 'addNewRecordButton')
    add_new_record_button.click

    profile_button = @driver.find_element(css: '.show #item-3')
    scroll_past_element(profile_button)
    profile_button.click

    profile_element = @driver.find_element(id: 'books-wrapper')
    @wait.until { profile_element.displayed? }

    @driver.find_element(id: 'see-book-Git Pocket Guide').click
    first_row_title_text = @driver.find_element(css: '#title-wrapper #userName-value').text
    assert_equal(book_name, first_row_title_text)

    @driver.navigate.back
    @driver.find_element(id: 'delete-record-undefined').click
  end

  def teardown
    @driver.quit
  end
end

File: src/test/C#/DemoQATest.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Interactions;
using OpenQA.Selenium.Support.UI;
using System;
using Xunit;

public class DemoQATest
{
    private void ScrollPastElement(IWebDriver driver, IWebElement element)
    {
        new Actions(driver).MoveToElement(element).Perform();
    }

    // [Fact]
    public void Test()
    {
        string username = "research_paper";
        string password = "1tq52rw43E!";

        ChromeOptions options = new ChromeOptions();
        //options.AddArgument("--headless=new");
        IWebDriver driver = new ChromeDriver(options);
        driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(4);
        driver.Manage().Window.Size = new System.Drawing.Size(1920, 1080);
        WebDriverWait wait = new WebDriverWait(driver, TimeSpan.FromSeconds(10));

        try
        {
            driver.Navigate().GoToUrl("https://demoqa.com/login");
            driver.FindElement(By.Id("userName")).SendKeys(username);
            driver.FindElement(By.Id("password")).SendKeys(password);
            driver.FindElement(By.Id("login")).Click();

            var storeButton = driver.FindElement(By.Id("gotoStore"));
            ScrollPastElement(driver, storeButton);
            storeButton.Click();

            driver.FindElement(By.Id("see-book-Git Pocket Guide")).Click();
            string bookName = driver.FindElement(By.CssSelector("#title-wrapper #userName-value")).Text;

            var addNewRecordButton = driver.FindElement(By.Id("addNewRecordButton"));
            addNewRecordButton.Click();

            var profileButton = driver.FindElement(By.CssSelector(".show #item-3"));
            ScrollPastElement(driver, profileButton);
            profileButton.Click();

            var profileElement = driver.FindElement(By.Id("books-wrapper"));
            wait.Until(driver => driver.FindElement(By.Id("books-wrapper")).Displayed);

            driver.FindElement(By.Id("see-book-Git Pocket Guide")).Click();
            string firstRowTitleText = driver.FindElement(By.CssSelector("#title-wrapper #userName-value")).Text;
            Assert.Equal(firstRowTitleText, bookName);

            driver.Navigate().Back();
            driver.FindElement(By.Id("delete-record-undefined")).Click();
        }
        finally
        {
            driver.Quit();
        }
    }
}
